pkg/proxy/service.go

```diff
@@ -0,0 +1,193 @@
+	for svcPortName, info := range serviceMap {
+		if info.HealthCheckNodePort() != 0 {
+			result.HCServiceNodePorts[svcPortName.NamespacedName] = uint16(info.HealthCheckNodePort())
+		}
+	}
+
+	return result
+}
+
+// apply the changes to ServiceMap and update the stale udp cluster IP set. The `UDPStaleClusterIP` argument is passed in to store the
+// udp protocol service cluster ip when service is deleted from the ServiceMap.
+func (serviceMap *ServiceMap) apply(changes *ServiceChangeMap, UDPStaleClusterIP sets.String) {
+	changes.lock.Lock()
+	defer changes.lock.Unlock()
+	for _, change := range changes.items {
+		existingPorts := serviceMap.Merge(change.current)
+		serviceMap.Unmerge(change.previous, existingPorts, UDPStaleClusterIP)
+	}
+	// clear changes after applying them to ServiceMap.
+	changes.items = make(map[types.NamespacedName]*serviceChange)
+	return
+}
+
+// Merge adds other ServiceMap's elements to current ServiceMap.  If exists, update the existing one, otherwise create new one.
+// It returns a string type set which stores all the newly merged services' identifier - ServicePortName.String() to help users
+// tell if a service is deleted or updated.  The returned value is one of the arguments of Unmerge().  For example,
+// - assume serviceMap A{} want to merge serviceMap B{{"ns", "cluster-ip", "http"}: {"172.16.55.10", 1234, "TCP"}}, then it will
+//   - change serviceMap A to be {{"ns", "cluster-ip", "http"}: {"172.16.55.10", 1234, "TCP"}} and
+//   - produce returned value {"ns/cluster-ip:http"}
+// - assume serviceMap A{{"ns", "cluster-ip", "http"}: {"172.16.55.10", 345, "UDP"}, {"ns", "node-port", "https"}: {"172.16.55.10", 443, "TCP"}} want to merge serviceMap B{{"ns", "cluster-ip", "http"}: {"172.16.55.10", 1234, "TCP"}}, then it will
+//   - change serviceMap A to be {{"ns", "cluster-ip", "http"}: {"172.16.55.10", 1234, "TCP"}, {"ns", "node-port", "https"}: {"172.16.55.10", 443, "TCP"}} and
+//   - produce returned value {"ns/cluster-ip:http"}
+func (sm *ServiceMap) Merge(other ServiceMap) sets.String {
+	// existingPorts is going to store all the unique newly merged services' identifier.
+	existingPorts := sets.NewString()
+	for svcPortName, info := range other {
+		// Take ServicePortName.String() as the newly merged service's identifier and put it into existingPorts.
+		existingPorts.Insert(svcPortName.String())
+		_, exists := (*sm)[svcPortName]
+		if !exists {
+			glog.V(1).Infof("Adding new service port %q at %s", svcPortName, info.String())
+		} else {
+			glog.V(1).Infof("Updating existing service port %q at %s", svcPortName, info.String())
+		}
+		(*sm)[svcPortName] = info
+	}
+	return existingPorts
+}
+
+// Unmerge deletes all other ServiceMap's elements from current ServiceMap.  The returned value of Merge() is the `existingPorts` argument of Unmerge().
+// If a service's identifier exists in `existingPorts`, it means it's an `Update` event and will skip delete, otherwise it's a `Delete` event and will
```

Comment:
existingPorts is an example of a weird API that is OK privately but less so publicly.

Can we change this API so that the caller has to pre-filter?  It looks like it is only called from one place, so just adding a ServiceMap.Filter(ports sets.String) or even changing merge to return a ServiceMap instead of string set?

I could take this as a followup if needed, but please think about it.  It's just a bad API to expose.  Weird.

Reply:
Okay, I will think about it.

Reply:
I fixed it by adding a `ServiceMap.filter(ports sets.String)` before calling unmerge. 

BTW, I think we take `merge()`, `unmerge()` and `filter()` as private.

Please take a look to see if you are happy with these changes :)